### PR TITLE
Threaded neural strat. This time it works :)

### PR DESCRIPTION
**Pull request by ghost**
Threaded neural strat. This time it works :)

Thanks for everyone and all for inspiration in creating this.
Thanks to Harvard for the code.
Thanks to NodeJS for including clustering in nodejs8!
Thanks to GNU parallel for main inspiration.

Thanks to zenbot, deviavir, carlos8f, the weather man, VertigoRay, OpticTrollz, BruceTus, BobTheSiaCoinGuy, that Japanese guy from napajs that I feel sorry for. Thanks for keeping me sane through coding these strategies. Also thanks for my main inspiration, weather prediction and sliding max-min algorithm for developing the signaling I call sliding 2/3ds averaging prediction.

Please comment, and review as I'd love to have this PR.

**Thanks to JensvdHeydt for helping with the speed issue and net creation too! :)**

Cluster module is native in nodejs8. Please dev test this for me guys! :) Predict some neural strategies!!

### CODE REVIEW

**Comment by JensvdHeydt** [comment]
@TheRoboKitten Thx for taking about 70% of the code that I rewrote for you or better: Thanks for using my coding without giving any form of credit.

Another thing: you are using the cluster feature but it looks like all your forks do the same work in the neural strategy. Where in the code does it divide the work between all those forks it generates?

File: conf-sample.js [hunk 1]
```diff
@@ -11,9 +11,9 @@ c.mongo.password = null
 c.mongo.replicaSet = null
 
 // default selector. only used if omitting [selector] argument from a command.
-c.selector = 'gdax.BTC-USD'
+c.selector = 'gdax.LTC-USD'
```
Comment (DeviaVir) [inline conf-sample.js]: I think BTC is a bit more common.

File: conf-sample.js [hunk 2]
```diff
@@ -11,9 +11,9 @@ c.mongo.password = null
...
 // name of default trade strategy
-c.strategy = 'trend_ema'
+c.strategy = 'trendline'
```
Comment (DeviaVir) [inline conf-sample.js]: Are you that confident that it should become the new default?

File: extensions/strategies/neural/strategy.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
+var convnetjs = require('convnetjs')
+var z = require('zero-fill')
+var stats = require('stats-lite')
+var n = require('numbro')
+var math = require('mathjs')
+const cluster = require('cluster');
+const numCPUs = require('os').cpus().length;
+// the below line starts you at 0 threads
+global.forks = 0
+// the beow line is for calculating the last mean vs the now mean.
```
Comment (DeviaVir) [inline extensions/strategies/neural/strategy.js]: typo: `below`

File: package.json [hunk 4]
```diff
@@ -19,19 +19,21 @@
...
     "bl": "^1.2.1",
     "btce": "^0.4.2",
-    "ccxt": "^1.10.119",
     "cexio-api-node": "^1.0.4",
     "cliff": "^0.1.10",
     "codemap": "^1.3.1",
     "colors": "^1.1.2",
     "commander": "^2.9.0",
+    "convnetjs": "0.3.0",
+    "ccxt": "^1.4.77",
```
Comment (DeviaVir) [inline package.json]: You downgraded this from 1.10 to 1.4

**Review by DeviaVir — CHANGES_REQUESTED** [review]

**Comment by JensvdHeydt** [comment]
Strategy "trendline". Something is broken. Thats trading in paper mode

2017-11-19 11:31:55  6649.75 BTC-EUR    0.00%        0   61 ++        NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   -0.01%
2017-11-19 11:32:00  6645.85 BTC-EUR   -0.06%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:32:20  6645.84 BTC-EUR    0.00%        1     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:32:25  6645.85 BTC-EUR   +0.00%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:32:35  6645.84 BTC-EUR    0.00%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:32:40  6645.85 BTC-EUR   +0.00%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:32:50  6645.85 BTC-EUR   +0.00%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:33:00  6645.84 BTC-EUR    0.00%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:33:05  6645.84 BTC-EUR   +0.00%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:33:10  6645.85 BTC-EUR   +0.00%        0     - 47       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:33:40  6645.80 BTC-EUR    0.00%        0     - 46       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:33:45  6645.51 BTC-EUR    0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:33:50  6645.50 BTC-EUR    0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:33:55  6645.50 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:05  6645.51 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:10  6645.51 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:15  6645.51 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:20  6645.51 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:25  6645.51 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:35  6645.50 BTC-EUR    0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:40  6645.51 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%
2017-11-19 11:34:45  6645.51 BTC-EUR   +0.00%        0    -- 45       NaN     sell  0.00000 BTC  1000.00 EUR   +0.00%   +0.05%

**Comment by ghost** [comment]
> Strategy "trendline". Something is broken. Thats trading in paper mode
Increase min_periods to something very very larger than the avgtrades/long part of the trendline.
@ 1 s period not enough trades in s.lookback[s.options.min_periods] when it starts calculating and then will error at 1000 average/long trades & 1000 min_periods.

Personally try run at min_periods 4000 and then average 105 and short 5 with .1 markup. 👍 

> @TheRoboKitten Thx for taking about 70% of the code that I rewrote for you or better: Thanks for using my coding without giving any form of credit.
> 
> Another thing: you are using the cluster feature but it looks like all your forks do the same work in the neural strategy. Where in the code does it divide the work between all those forks it generates?

This needs to be fixed I admit, I was going to give you credit but I am bad with names.

I am considering removing the threaded code, or simply, set your threads to 1 for now. I'll set as default. :) 🔢

**Comment by ghost** [comment]
If you want the PR take it or leave it. I have no future plans to support zenbot due to nodejs limitaitons.

**Comment by vitorego** [comment]
I´m using the trendline strategy and it is good. For BTC-EUR on gdax need to change the default options, but is having good results on paper mode.

**Comment by JensvdHeydt** [comment]
@vitorego I think the trendline strategy is interesting. It just needs some polishing because it gave me undefined values as trade statistic and that's not very trustworthy :)

**Comment by halvors** [comment]
@vitorego What options where you using?

**Comment by vitorego** [comment]
@halvors:
`./zenbot.sh trade --strategy=trendline --min_periods=200 --avgpoints=100 --lastpoints=10 --period=30m gdax.BTC-EUR.default"`

**Comment by vitorego** [comment]
@JensvdHeydt I also got that error on bitfinex using pair IOT-USD. I used the same options as I used for gdax pair BTC-EUR:

`1786  ./zenbot.sh sim --verbose --strategy=trendline --min_periods=200 --avgpoints=100 --lastpoints=10 --period=30m --days=15 bitfinex.IOT-USD`

**Comment by mkuendig** [comment]
have this PR running on my zenbot but no actions after 36 hours:

./zenbot.sh trade poloniex.STR-BTC --strategy neural

It has started to trade now. Had to reboot the container. However, now it trades too often on very tiny profits and too often after exchange fee the trade is negative.

**Review by DeviaVir — COMMENTED** [review]
LGTM now, I'll merge so we get more eyes on these strats.

**Review by DeviaVir — APPROVED** [review]

**Pull request merged**